x/profile/internal/keeper/keeper.go

```diff
@@ -0,0 +1,132 @@
+	return Keeper{
+		StoreKey: storeKey,
+		Cdc:      cdc,
+	}
+}
+
+// AssociateMonikerWithAddress save the relation of moniker and address on chain
+func (k Keeper) AssociateMonikerWithAddress(ctx sdk.Context, moniker string, address sdk.AccAddress) {
+	store := ctx.KVStore(k.StoreKey)
+	key := types.MonikerStoreKey(moniker)
+	store.Set(key, k.Cdc.MustMarshalBinaryBare(&address))
+}
+
+// GetMonikerRelatedAddress returns the address associated to the given moniker or nil if it not exists
+func (k Keeper) GetMonikerRelatedAddress(ctx sdk.Context, moniker string) (addr sdk.AccAddress) {
+	store := ctx.KVStore(k.StoreKey)
+	bz := store.Get(types.MonikerStoreKey(moniker))
+	if bz == nil {
+		return nil
+	}
+	k.Cdc.MustUnmarshalBinaryBare(bz, &addr)
+	return addr
+}
+
+// GetMonikerFromAddress returns the moniker associated with the given address or an empty string if no moniker exists
+func (k Keeper) GetMonikerFromAddress(ctx sdk.Context, addr sdk.AccAddress) (moniker string) {
+	store := ctx.KVStore(k.StoreKey)
+	it := sdk.KVStorePrefixIterator(store, types.MonikerStorePrefix)
+	defer it.Close()
+
+	for ; it.Valid(); it.Next() {
+		var acc sdk.AccAddress
+		k.Cdc.MustUnmarshalBinaryBare(it.Value(), &acc)
+		if acc.Equals(addr) {
+			return string(bytes.TrimPrefix(it.Key(), types.MonikerStorePrefix))
+		}
+	}
+
+	return ""
+}
+
+// DeleteMonikerAddressAssociation delete the given moniker association with an address
+func (k Keeper) DeleteMonikerAddressAssociation(ctx sdk.Context, moniker string) {
+	store := ctx.KVStore(k.StoreKey)
+	key := types.MonikerStoreKey(moniker)
+	store.Delete(key)
+}
+
+// SaveProfile allows to save the given account inside the current context.
+// It assumes that the given account has already been validated.
+// It returns an error if an account with the same moniker from a different creator already exists
+func (k Keeper) SaveProfile(ctx sdk.Context, profile types.Profile) error {
+	store := ctx.KVStore(k.StoreKey)
+
+	key := types.ProfileStoreKey(profile.Creator.String())
+
+	moniker := k.GetMonikerFromAddress(ctx, profile.Creator)
+	k.DeleteMonikerAddressAssociation(ctx, moniker)
+
+	if addr := k.GetMonikerRelatedAddress(ctx, profile.Moniker); addr != nil && !addr.Equals(profile.Creator) {
+		return fmt.Errorf("an account with moniker: %s has already been created", profile.Moniker)
+	}
+
+	k.AssociateMonikerWithAddress(ctx, profile.Moniker, profile.Creator)
+	store.Set(key, k.Cdc.MustMarshalBinaryBare(&profile))
+
+	return nil
+}
+
+// DeleteProfile allows to delete an account associated with the given address inside the current context.
+// It assumes that the address-related account exists.
+// nolint: interfacer
+func (k Keeper) DeleteProfile(ctx sdk.Context, address sdk.AccAddress, moniker string) {
+	store := ctx.KVStore(k.StoreKey)
+	key := types.ProfileStoreKey(address.String())
+	store.Delete(key)
+	k.DeleteMonikerAddressAssociation(ctx, moniker)
```

Comment:
Not a big fan of uselss variable allocations. I think we can remove `key` here

```suggestion
	store := ctx.KVStore(k.StoreKey)
	store.Delete(types.ProfileStoreKey(address.String()))
	k.DeleteMonikerAddressAssociation(ctx, moniker)
```